elephant/spike_train_generation.py

```diff
@@ -333,46 +344,694 @@ def peak_detection(signal, threshold=0.0 * pq.mV, sign='above',
+    rate : pq.Quantity
+        The constant firing rate.
+    cv : float
+        The expected coefficient of variation.
+    t_start : pq.Quantity, optional
+        The start of the spike train.
+        Default: 0.*pq.s
+    t_stop : pq.Quantity, optional
+        The end of the spike train.
+        Default: 1.*pq.s
+    equilibrium : bool, optional
+        Generate an equilibrium or an ordinary renewal process.
+        Default: True
+
+    Examples
+    --------
+    >>> import quantities as pq
+    >>> spiketrain = StationaryInverseGaussianProcess(
+    ...     rate=50*pq.Hz, cv=2.0, t_start=0*pq.ms,
+    ...     t_stop=1000*pq.ms).generate_spiketrain()
+    >>> spiketrain_array = StationaryInverseGaussianProcess(
+    ...     rate=20*pq.Hz, cv=5.0, t_start=5000*pq.ms,
+    ...     t_stop=10000*pq.ms).generate_spiketrain(as_array=True)
+    """
+    def __init__(
+            self,
+            rate: pq.Quantity,
+            cv: float,
+            t_start: pq.Quantity = 0. * pq.s,
+            t_stop: pq.Quantity = 1.*pq.s,
+            equilibrium: bool = True
+    ):
+        super().__init__(
+            rate=rate, t_start=t_start, t_stop=t_stop, equilibrium=equilibrium)
+        self._cv = cv
+        if self.n_expected_spikes > 0:
+            self.isi_generator = stats.invgauss(
+                mu=cv**2, scale=1./(self.rate*cv**2))
+
+    @property
+    def expected_cv(self):
+        """
+        The expected coefficient of variation given the ISI distribution.
+        """
+        return self._cv
+
+
+class RateModulatedProcess(AbstractPointProcess):
+    """
+    Abstract rate-modulated process to subclass from.
+
+    Parameters
+    ----------
+    rate_signal : neo.AnalogSignal
+        A `neo.AnalogSignal` representing the rate profile evolving over
+        time.Its values have all to be `>=0`. The generated spike trains
+        will have `t_start = rate.t_start` and `t_stop = rate.t_stop`
+
+    Raises
+    ------
+    ValueError
+        If `rate_signal` is not a neo AnalogSignal
+        If `rate_signal` contains a negative value.
+        If `rate_signal` is empty.
+    """
+    process_operational_time: RenewalProcess
+
+    def __init__(self, rate_signal: neo.AnalogSignal):
+
+        if not isinstance(rate_signal, neo.AnalogSignal):
+            raise ValueError(
+                f'rate_signal should be of type neo.AnalogSignal.'
+                f' Currently it is of type: {type(rate_signal)}')
+        if len(rate_signal) == 0:
+            raise ValueError('rate_signal can not be empty.')
+        if not all(rate_signal >= 0.):
+            raise ValueError(
+                'All elements of rate_signal should be positive.')
+
+        super().__init__(
+            t_start=rate_signal.t_start, t_stop=rate_signal.t_stop)
+
+        self.rate_signal = rate_signal
+
+        self.mean_rate = np.mean(rate_signal.rescale(1./self.units).magnitude)
+
+        if self.mean_rate == 0.:
+            return
```

Comment:
The idea is to get out of the init function, like a break in a for-loop. Since the following parameters can't be calculated for the case with zero rate. I added a comment.